homework/tai_nguyen/lesson7.2/JS_questions.md

```diff
@@ -0,0 +1,136 @@
+**Question 1:** Write a function printNumbers(from, to) that outputs a number every second, starting from from and ending with to. Make two variants of the solution.
+
+```typescript
+// Using setInterval
+function printNumber(from: number, to: number) {
+  let intervalId: number;
+
+  intervalId = setInterval(() => {
+    if (from <= to) {
+      console.log(from);
+    }
+
+    if (from === to) {
+      clearInterval(intervalId);
+    }
+
+    from++;
+  }, 1 * 1000);
+}
+
+// Using nested setTimeout
+function printNumber(from: number, to: number) {
+  function schedule(x: number) {
+    if (x > to) {
+      return;
+    }
+
+    setTimeout(() => {
+      console.log(x);
+      schedule(x + 1);
+    }, 1 * 1000);
+  }
+
+  schedule(from);
+}
+```
+
+**Question 2:** Use the reduce method in combination with the concat method to “flatten” an array of arrays into a single array that has all the elements of the original arrays.
+
+```typescript
+function flatten(array: any[][]) {
+  return array.reduce<any[]>((acc, item) => acc.concat(item), []);
+}
+
+let arrays = [[1, 2, 3], [4, 5], [6]];
+console.log(flatten(arrays));
+```
+
+**Question 3:** Write a higher-order function loop that provides something like a for loop statement. It takes a value, a test function, an update function, and a body function
+
+```typescript
+type TestFunc = (x: number) => boolean;
```

Comment:
requirement of this question must use higher-order function

Reply:
@trongtai37 need to update this ?

Reply:
Hi @thaycacac, my loop function is already a higher-order function because it takes other functions as arguments.
Can you explain your point more?

Reply:
@trongtai37  higher-order function must be returns a function as its result.

Reply:
I see the definition as below:
> A “higher-order function” is a function that accepts functions as parameters and/or returns a function.
Anw, Can you help me with writing `loop` function as return the function?

Reply:
```js
function loop(value, test, update, execute) {
  if (!test(value)) return;
  execute(value);
  return loop(update(value), test, update, execute);
}

loop(3, n => n > 0, n => n - 1, console.log);
```

Reply:
Thanks @thaycacac,
It's a recursive approach and `loop` function return `undefined` or  result of `loop` func.
I tried to log the result of `console.log(loop(3, n => n > 0, n => n - 1, console.log));` and it always return `undefined`